sql-cli/src/sql_cli/sql_directory_parser.py

```diff
@@ -0,0 +1,132 @@
+    def get_content(self) -> str:
+        """
+        Get the content of the SQL file.
+
+        :returns: the SQL file's content as string.
+        """
+        with open(self.path) as file:
+            return file.read()
+
+    def get_variable_name(self) -> str:
+        """
+        Get the variable name used as a unique identifier as a python variable.
+
+        :returns: the file name without suffix.
+        """
+        if self.has_directory():
+            return f"{'_'.join(self.get_directories())}_{self.path.stem}"
+        return self.path.stem
+
+    def get_output_table_name(self) -> str:
+        """
+        Get the output table name.
+
+        :returns: the output table name.
+        """
+        return self.path.stem
+
+    def get_output_schema_name(self) -> str | None:
+        """
+        Get the output schema name.
+
+        :returns: the output schema name.
+        """
+        if self.has_directory():
+            return self.path.parent.name
+        return None
+
+    def get_parameters(self) -> set[str]:
+        """
+        Get all parameters used for parameterized sql queries.
+
+        :returns: declared parameters for the sql query.
+        """
+        return find_template_variables(self.path)  # type: ignore
+
+    def get_directories(self) -> list[str]:
+        """
+        Get the parents directory names.
+
+        :returns: the parents directory names.
+        """
+        parents = []
+        for parent in self.path.parents:
+            if parent == self.root_dir:
+                break
+            parents.append(parent.name)
+        parents.reverse()
+        return parents
+
+    def has_directory(self) -> bool:
+        """
+        Check if sql file is in a sub directory.
+
+        :returns: True if it's not in the parent root directory.
+        """
+        return self.path.parent != self.root_dir
+
+
+class SqlDirectory:
+    """
+    A directory of SQL files.
+
+    :param path: The path to the directory containing sql files.
+    """
+
+    def __init__(self, path: Path) -> None:
+        self.path = path
+
+    def get_variable_name(self) -> str:
+        """
+        Get the variable name used as a unique identifier as a python variable.
+
+        :returns: the directory name.
+        """
+        return self.path.name
+
+
+def get_sql_directories(directory: Path) -> set[SqlDirectory]:
+    """
+    Get all sql directories within a directory.
+
+    :param directory: The directory look in for sql directories.
+
+    :returns: the sql directories found in the directory.
+    """
+    return {
+        SqlDirectory(path=child) for child in directory.rglob("*") if child.is_dir()
+    }
+
+
+def get_sql_files(directory: Path) -> set[SqlFile]:
+    """
+    Get all sql files within a directory.
+
+    :param directory: The directory look in for sql files.
+
+    :returns: the sql files found in the directory.
+    """
+    return {
+        SqlFile(root_dir=directory, path=child)
```

Comment:
Should we create a taskgroup based on the directory name for nested directories?

Reply:
Is this a feature request? Or a rhetorical question? 😅 The current implementation via `rglob("*.sql")` does a recursive traversal. So we find **all** directories.

I think yes. I think it makes sense to reflect the directory structure in the database as schemas.